src/plugins/run/annotations.ts

```diff
@@ -0,0 +1,203 @@
+export const smdAnnotations: unified.Attacher = function () {
+  return function transform($root) {
+    const root = $root as MDAST.Root;
+    const nodes = root.children;
+
+    const processed: MDAST.Content[] = [];
+
+    let inTab: boolean = false;
+    let skipNext: boolean = false;
+
+    // temporary variable for storing current tabContainer
+    let tabPlaceholder: MDAST.Tabs = {
+      type: 'tabs',
+      data: {
+        hName: 'tabs',
+      },
+      children: [
+        {
+          type: 'tab',
+          data: {
+            hName: 'tab',
+          },
+          children: [],
+        },
+      ],
+    };
+
+    for (const i in nodes) {
+      if (!nodes[i]) continue;
+
+      if (skipNext) {
+        skipNext = false;
+        continue;
+      }
+
+      // this node
+      const node = nodes[i];
+
+      // next node
+      const next = nodes[+i + 1] ? nodes[+i + 1] : null;
+
+      // collect annotations, if this is an html node
+      const anno = captureAnnotations(node);
+
+      if ('type' in anno) {
+        const { type } = anno;
+
+        if (type === 'tab') {
+          const { children } = tabPlaceholder;
+
+          if (inTab && tabPlaceholder) {
+            // already inside of a tab, so this is a new one
+            children.push({
+              type: 'tab',
+              data: {
+                hName: 'tab',
+              },
+              children: [],
+            });
+          } else {
+            // not inside a tab already
+            inTab = true;
+          }
+
+          // set annotations if present
+          if (Object.keys(anno).length > 0) {
+            Object.assign(children[children.length - 1].data, {
+              hProperties: anno,
+            });
+          }
+
+          tabPlaceholder.children = children;
+
+          continue;
+        } else if (type === 'tab-end') {
+          // finalize tabContainer
+          processed.push(tabPlaceholder);
+
+          // reset tabPlaceholder
+          inTab = false;
+          tabPlaceholder = {
+            type: 'tabs',
+            data: {
+              hName: 'tabs',
+            },
+            children: [
+              {
+                type: 'tab',
+                data: {
+                  hName: 'tab',
+                },
+                children: [],
+              },
+            ],
+          };
+
+          continue;
+        }
+      }
+
+      if (inTab) {
+        // if we're in a tab, push this node as a child of the last tab
+        const size = tabPlaceholder.children.length;
+        if (tabPlaceholder.children[size - 1]) {
+          (tabPlaceholder.children[size - 1].children as MDAST.Content[]).push(processNode(node, anno));
+        }
+      } else if (Object.keys(anno).length > 0 && next) {
+        // annotations apply to next node, process next node now and skip next iteration
+        processed.push(processNode(next, anno));
+        skipNext = true;
```

Comment:
or actually, here is a nice trick if we use `for of`.

```ts
const entries = nodes.entries()[Symbol.iterator]();
for (const [i, node] of entries) {
  
```

```suggestion
        entries.next(); // boom! skipped. no need for that if above
```

Reply:
Nice, works! Neat trick - I learned something today.